### PR TITLE
[DTM0] DTM0 patch for main (review board)

### PR DESCRIPTION
**Pull request by ivan-alekhin**
A PR for overall review of the DTM0 (the happy-path part). 

The branch (`dtm0-main2main`) will be the place to submit the corresponding fixes (coding style issues and so on).

The PR should remain in the draft state. Later on we will submit a separate PR to be reviewed by the upstream gatekeepers.

UPDATE: As per discussions with @just-now , we are merging this PR into dtm0-main. After that, a separate "official" `dtm0-main -> main` PR will be open for review.

### CODE REVIEW

File: conf/objs/service.c [hunk 1]
```diff
@@ -75,9 +82,17 @@ service_encode(struct m0_confx_obj *dest, const struct m0_conf_obj *src)
...
 		return M0_ERR(-ENOMEM);
+
+	rc = m0_bufs_from_strings(&d->xs_params, s->cs_params);
+	if (rc != 0) {
+		m0_bufs_free(&d->xs_endpoints);
+		return M0_ERR(-ENOMEM);
+	}
+
 	rc = arrfid_from_dir(&d->xs_sdevs, s->cs_sdevs);
 	if (rc != 0)
```
Comment (just-now) [inline conf/objs/service.c:93]: bug:
if () {}

File: dix/req.c [hunk 2]
```diff
@@ -1678,11 +1711,22 @@ static bool dix_cas_rop_clink_cb(struct m0_clink *cl)
...
-		if (rop->dg_completed_nr == rop->dg_cas_reqs_nr) {
-			rop->dg_ast.sa_cb = dix_rop_completed;
-			rop->dg_ast.sa_datum = dreq;
+
+		if (dreq->dr_dtx != NULL) {
+			rop->dg_ast.sa_cb = dix_rop_one_completed;
+			rop->dg_ast.sa_datum = crop;
+			M0_ASSERT(dix_req_smgrp(dreq) ==
+				  dreq->dr_dtx->tx_dtx->dd_sm.sm_grp);
 			m0_sm_ast_post(dix_req_smgrp(dreq), &rop->dg_ast);
```
Comment (just-now) [inline dix/req.c:1720]: move outside.

File: dix/req.c [hunk 3]
```diff
@@ -1678,11 +1711,22 @@ static bool dix_cas_rop_clink_cb(struct m0_clink *cl)
 		rop = crop->crp_parent->dr_rop;
 		rop->dg_completed_nr++;
 		M0_PRE(rop->dg_completed_nr <= rop->dg_cas_reqs_nr);
-		if (rop->dg_completed_nr == rop->dg_cas_reqs_nr) {
-			rop->dg_ast.sa_cb = dix_rop_completed;
-			rop->dg_ast.sa_datum = dreq;
+
+		if (dreq->dr_dtx != NULL) {
+			rop->dg_ast.sa_cb = dix_rop_one_completed;
+			rop->dg_ast.sa_datum = crop;
```
Comment (just-now) [inline dix/req.c:1717]: please use rop-> dg_ast = {.sa_cb = dix_rop_one_completed, sa_datum = crop}.
rationale:
don't need to look at other fields of ast, as they will be nulified.

File: dix/req.c [hunk 4]
```diff
@@ -1755,14 +1800,26 @@ static int dix_cas_rops_send(struct m0_dix_req *req)
...
 		}
+		i++;
 	} m0_tl_endfor;
 
 	M0_LOG(M0_DEBUG, "Processing dix_req %p rop=%p: dg_cas_reqs_nr=%"PRIu64,
 				  req, rop, rop->dg_cas_reqs_nr);
 	if (rop->dg_cas_reqs_nr == 0)
 		return M0_ERR(-EFAULT);
+
+	if (req->dr_dtx) {
```
Comment (just-now) [inline dix/req.c:1817]: != NULL

File: dix/req.c [hunk 5]
```diff
@@ -1791,7 +1848,11 @@ static uint32_t dix_rop_tgt_iter_max(struct m0_dix_req    *req,
 		 */
 		return m0_dix_liter_P(iter);
 	else
-		return m0_dix_liter_N(iter) + 2 * m0_dix_liter_K(iter);
+		/* Skip spares in ENABLE_DTM0 */
+		if (ENABLE_DTM0)
```
Comment (just-now) [inline dix/req.c:1852]: return ENABLE_DTM0  ? m0_dix_liter_N(iter) + m0_dix_liter_K(iter)  : m0_dix_liter_N(iter) + 2 * m0_dix_liter_K(iter);

File: dtm0/service.c [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+
+static void dtm0_service__fini(struct m0_dtm0_service *s)
+{
+	m0_dtm0_clk_src_fini(&s->dos_clk_src);
+	m0_dtm0_dtx_domain_fini();
+	dopr_tlist_fini(&s->dos_processes);
+	m0_dtm0_service_bob_fini(s);
+}
+
+static struct dtm0_process *dtm0_service_process__lookup(struct m0_reqh_service *reqh_dtm0_svc,
```
Comment (just-now) [inline dtm0/service.c:153]: looks too bloody...


let's try this one:
static struct dtm0_process *
dtm0_service_process__lookup(struct m0_reqh_service *reqh_dtm0_svc,
			     const struct m0_fid *remote_dtm0)
{
	return m0_tl_find(dopr, proc, &to_dtm(reqh_dtm0_svc)->dos_processes,
			  m0_fid_eq(&proc->dop_rserv_fid, remote_dtm0));
}

File: dtm0/service.c [hunk 7]
```diff
@@ -0,0 +1,625 @@
...
+	struct m0_rpc_machine *mach =
+		m0_reqh_rpc_mach_tlist_head(&s->rs_reqh->rh_rpc_machines);
+	int rc;
+
+	process = dtm0_service_process__lookup(s, remote_srv);
+	if (process == NULL)
+		return M0_RC(-ENOENT);
+
+	rc = m0_rpc_link_init(&process->dop_rlink, mach, remote_srv,
+			      remote_ep, 10);
```
Comment (just-now) [inline dtm0/service.c:183]: why 10? put a comment.

File: dtm0/service.c [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+	M0_ASSERT(rc == 0);
```
Comment (just-now) [inline dtm0/service.c:184]: why assert? put a comment

File: dtm0/service.c [hunk 9]
```diff
@@ -0,0 +1,625 @@
...
+	M0_LOG(M0_DEBUG,
+	       " async=%d"
+	       " dtm0="FID_F
+	       " remote_srv="FID_F,
+	       !!async,
+	       FID_P(&s->rs_service_fid),
+	       FID_P(remote_srv));
+	M0_LOG(M0_DEBUG, "rep=%s", remote_ep);
+
+	if (!async)
```
Comment (just-now) [inline dtm0/service.c:195]: rc = !async ? m0_rpc_link_connect_sync(&process->dop_rlink,
					       M0_TIME_NEVER) :
		m0_rpc_link_connect_async(&process->dop_rlink,
					  M0_TIME_NEVER,
					  &process->dop_service_connect_clink);

File: dtm0/service.c [hunk 10]
```diff
@@ -0,0 +1,625 @@
...
+	return M0_RC(rc);
+}
+
+M0_INTERNAL int m0_dtm0_service_process_disconnect(struct m0_reqh_service *s,
+						   struct m0_fid *remote_srv)
+{
+	int rc;
+	struct dtm0_process *process = NULL;
+
+	M0_LOG(M0_DEBUG,
```
Comment (just-now) [inline dtm0/service.c:212]: please put it into a single line

File: dtm0/service.c [hunk 11]
```diff
@@ -0,0 +1,625 @@
...
+	rc = m0_rpc_link_disconnect_sync(&process->dop_rlink,
+					 m0_time_from_now(5, 0));
+
+	M0_ASSERT(rc == 0 || rc == -ETIMEDOUT);
+	if (rc == -ETIMEDOUT)
+		M0_LOG(M0_WARN, "Disconnect timeout");
+
+	m0_rpc_link_fini(&process->dop_rlink);
+
+	return M0_RC(0);
```
Comment (just-now) [inline dtm0/service.c:232]: rc?

File: dtm0/service.c [hunk 12]
```diff
@@ -0,0 +1,625 @@
...
+
+	M0_ALLOC_PTR(s);
+	if (s == NULL)
+		return M0_ERR(-ENOMEM);
+
+	s->dos_generic.rs_type = stype;
+	s->dos_generic.rs_ops  = ops;
+	rc = dtm0_service__init(s);
+	if (rc != 0)
+		return rc;
```
Comment (just-now) [inline dtm0/service.c:262]: m0_rc

File: dtm0/service.c [hunk 13]
```diff
@@ -0,0 +1,625 @@
...
+
+	*service = &s->dos_generic;
+	return M0_RC(0);
```
Comment (just-now) [inline dtm0/service.c:265]: rc

File: dtm0/service.c [hunk 14]
```diff
@@ -0,0 +1,625 @@
...
+        M0_PRE(service != NULL);
+        return dtm_service__origin_fill(service) ?:
+		    dtm0_service__ha_subscribe(service);
+}
+
+static void dtm0_service_stop(struct m0_reqh_service *service)
+{
+	struct m0_dtm0_service *dtm0;
+
+        M0_PRE(service != NULL);
```
Comment (just-now) [inline dtm0/service.c:329]: spc

File: dtm0/service.c [hunk 15]
```diff
@@ -0,0 +1,625 @@
...
+	 * until TRANSIENT is received. TRANSIENT is just a
+	 * trigger for handling HA messages in this case.
+	 */
+	if (evented_proc_state != M0_NC_TRANSIENT &&
+	    !ready_to_process_ha_msgs && !m0_dtm0_in_ut()) {
+		M0_LOG(M0_DEBUG, "Is not ready to process HA messages");
+		goto out;
+	}
+
+	/* (M0_IN(obj->co_ha_state, (M0_NC_ONLINE, M0_NC_FAILED, M0_NC_TRANSIENT))) */
```
Comment (just-now) [inline dtm0/service.c:457]: remove

File: dtm0/service.c [hunk 16]
```diff
@@ -0,0 +1,625 @@
...
+	struct m0_reqh_service *dop_dtm0_service;
+	/**
+	 * Connect ast
+	 */
+	struct m0_sm_ast        dop_service_connect_ast;
+	struct m0_clink         dop_service_connect_clink;
+};
+
+M0_TL_DESCR_DEFINE(dopr, "dtm0_process", static, struct dtm0_process, dop_link,
+		   dop_magic, 0x8888888888888888, 0x7777777777777777);
```
Comment (just-now) [inline dtm0/service.c:112]: motr/magic.h waits for new children.

File: utils/trace/m0trace [hunk 17]
```diff
@@ -329,7 +329,7 @@ sub validate_subsys
         cas      formation  net     stob
         cm       ha         other   ut
         cob      ioservice  pool    xcode
-        conf     layout     rm
+        conf     layout     rm      dtm0
```
Comment (just-now) [inline utils/trace/m0trace:332]: 👍

File: motr/ut/idx_dix.c [hunk 18]
```diff
@@ -782,15 +782,282 @@ struct m0_client* st_get_instance()
...
+
+	rc = ut_suite_mt_idx_dix_fini();
+	m0_fi_disable("m0_dtm0_in_ut", "ut");
+	return rc;
+}
+
+/* Submits multiple M0 client (PUT|DEL) operations and then waits on EXECUTED,
+ * and then waits on STABLE.
+ */
+static void exec_then_stable(uint64_t nr, enum m0_idx_opcode opcode)
```
Comment (just-now) [inline motr/ut/idx_dix.c:897]: exec_then_stable && exec_until_stable are copy/pasted. please reduce the amount of duplication.

File: motr/ut/idx_dix.c [hunk 19]
```diff
@@ -782,15 +782,282 @@ struct m0_client* st_get_instance()
...
+	/* Delete the index */
+	rc = m0_entity_delete(&duc.duc_idx.in_entity, &op);
+	M0_UT_ASSERT(rc == 0);
+	m0_op_launch(&op, 1);
+	rc = m0_op_wait(op, M0_BITS(M0_OS_STABLE), WAIT_TIMEOUT);
+	M0_UT_ASSERT(rc == 0);
+	m0_op_fini(op);
+	m0_op_free(op);
+	m0_idx_fini(&duc.duc_idx);
+	op = NULL;
```
Comment (just-now) [inline motr/ut/idx_dix.c:872]: this is not needed.

File: motr/ut/idx_dix.c [hunk 20]
```diff
@@ -782,15 +782,282 @@ struct m0_client* st_get_instance()
...
+
+	/* Create the index */
+	rc = m0_entity_create(NULL, &idx->in_entity, &op);
+	M0_UT_ASSERT(rc == 0);
+	m0_op_launch(&op, 1);
+	rc = m0_op_wait(op, M0_BITS(M0_OS_STABLE), WAIT_TIMEOUT);
+	M0_UT_ASSERT(rc == 0);
+	m0_op_fini(op);
+	m0_op_free(op);
+	op = NULL;
```
Comment (just-now) [inline motr/ut/idx_dix.c:855]: this is not needed.

File: motr/ut/idx_dix.c [hunk 21]
```diff
@@ -782,15 +782,282 @@ struct m0_client* st_get_instance()
...
+	duc.duc_cli_svc_fid = cli_srv_fid;
+	duc.duc_srv_svc_fid = srv_dtm0_fid;
+	duc.duc_srv_svc = srv_srv;
+
+	return 0;
+}
+
+static void idx_setup(void)
+{
+	struct m0_op            *op = NULL;
```
Comment (just-now) [inline motr/ut/idx_dix.c:839]: such nullification is a sign of unexpected side effect inside the use of this structure. either put a comment why it's needed or avoid = NULL.

File: cas/service.c [hunk 22]
```diff
@@ -1087,6 +1119,30 @@ static int op_sync_wait(struct m0_fom *fom)
...
+		return M0_ERR(rc);
+
+	return rc;
+}
+
+M0_UNUSED static int cas_op_decode( void *obj, struct m0_buf *buf)
+{
+	int rc;
+
+	rc = m0_xcode_obj_dec_from_buf(&M0_XCODE_OBJ(m0_cas_op_xc, (void*)obj),
```
Comment (just-now) [inline cas/service.c:1138]: remove this code, it's not needed.

File: utils/m0setup [hunk 23]
```diff
@@ -229,7 +231,7 @@ parse_cli_options()
             -P|--pool-width)    pool_width=$2;        shift 2 ;;
             -N|--data-units)    data_units=$2;        shift 2 ;;
             -K|--parity-units)  parity_units=$2;      shift 2 ;;
-            --pool-width)       ipool_width=$2;       shift 2 ;;
+            --ipool-width)      ipool_width=$2;       shift 2 ;;
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply utils/m0setup:234]: It's a bug in previous implementation, I left the comment earlier.

File: dtm0/service.c (hunk 6 above)
Comment (sergey-shilov) [reply dtm0/service.c:153]: Done.

File: dtm0/service.c (hunk 8 above)
Comment (sergey-shilov) [reply dtm0/service.c:184]: Replaced with return.

File: dtm0/service.c (hunk 9 above)
Comment (sergey-shilov) [reply dtm0/service.c:195]: m0_rpc_link_connect_async() is void function.

File: dtm0/service.c (hunk 10 above)
Comment (sergey-shilov) [reply dtm0/service.c:212]: Done.

File: dtm0/service.c (hunk 7 above)
Comment (sergey-shilov) [reply dtm0/service.c:183]: Created corresponding macro.

File: dtm0/service.c (hunk 11 above)
Comment (sergey-shilov) [reply dtm0/service.c:232]: Done.

File: dtm0/service.c (hunk 12 above)
Comment (sergey-shilov) [reply dtm0/service.c:262]: Done.

File: dtm0/service.c (hunk 13 above)
Comment (sergey-shilov) [reply dtm0/service.c:265]: Done.

File: dtm0/service.c (hunk 14 above)
Comment (sergey-shilov) [reply dtm0/service.c:329]: Done.

File: dtm0/service.c (hunk 15 above)
Comment (sergey-shilov) [reply dtm0/service.c:457]: Done.

File: utils/m0setup (hunk 23 above)
[the remark this answers is not in the document]
Comment (just-now) [reply utils/m0setup:234]: 👍

**Review by just-now — APPROVED** [review]
looks good.

**Pull request merged**